File: data/code_pairs/fortran/882.f90
MODULE ppm_module_gmm_cpt
  IMPLICIT NONE
CONTAINS
  SUBROUTINE ppm_gmm_cpt_2ds()
    PRINT *, "2D Single Precision GMM Component"
  END SUBROUTINE ppm_gmm_cpt_2ds

  SUBROUTINE ppm_gmm_cpt_2dd()
    PRINT *, "2D Double Precision GMM Component"
  END SUBROUTINE ppm_gmm_cpt_2dd

  SUBROUTINE ppm_gmm_cpt_3ds()
    PRINT *, "3D Single Precision GMM Component"
  END SUBROUTINE ppm_gmm_cpt_3ds

  SUBROUTINE ppm_gmm_cpt_3dd()
    PRINT *, "3D Double Precision GMM Component"
  END SUBROUTINE ppm_gmm_cpt_3dd
END MODULE ppm_module_gmm_cpt

PROGRAM main
  USE ppm_module_gmm_cpt
  IMPLICIT NONE
  CALL ppm_gmm_cpt_2ds()
  CALL ppm_gmm_cpt_2dd()
  CALL ppm_gmm_cpt_3ds()
  CALL ppm_gmm_cpt_3dd()
END PROGRAM main

File: data/code_pairs/cpp/882.cpp
#include <iostream>

// Define constants for precision and dimension
#define __SINGLE_PRECISION         1
#define __DOUBLE_PRECISION         2
#define __2D                       3
#define __3D                       4

namespace ppm_module_gmm_cpt {

template<int DIM, int KIND>
class ppm_gmm_cpt {
public:
    static void process() {
        std::cout << "Generic GMM Component Processing" << std::endl;
    }
};

template<>
class ppm_gmm_cpt<__2D, __SINGLE_PRECISION> {
public:
    static void process() {
        std::cout << "2D Single Precision GMM Component" << std::endl;
    }
};

template<>
class ppm_gmm_cpt<__2D, __DOUBLE_PRECISION> {
public:
    static void process() {
        std::cout << "2D Double Precision GMM Component" << std::endl;
    }
};

template<>
class ppm_gmm_cpt<__3D, __SINGLE_PRECISION> {
public:
    static void process() {
        std::cout << "3D Single Precision GMM Component" << std::endl;
    }
};

template<>
class ppm_gmm_cpt<__3D, __DOUBLE_PRECISION> {
public:
    static void process() {
        std::cout << "3D Double Precision GMM Component" << std::endl;
    }
};

} // namespace ppm_module_gmm_cpt

int main() {
    ppm_module_gmm_cpt::ppm_gmm_cpt<__2D, __SINGLE_PRECISION>::process();
    ppm_module_gmm_cpt::ppm_gmm_cpt<__2D, __DOUBLE_PRECISION>::process();
    ppm_module_gmm_cpt::ppm_gmm_cpt<__3D, __SINGLE_PRECISION>::process();
    ppm_module_gmm_cpt::ppm_gmm_cpt<__3D, __DOUBLE_PRECISION>::process();

    return 0;
}